src/node_crypto.cc

```diff
@@ -1631,6 +1662,74 @@ void SSLWrap<Base>::IsInitFinished(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+inline int compar(const void* a, const void* b) {
+  return memcmp(a, b, CNNIC_WHITELIST_HASH_LEN);
+}
+
+
+inline int IsSelfSigned(X509* cert) {
+  return X509_NAME_cmp(X509_get_subject_name(cert),
+                       X509_get_issuer_name(cert)) == 0;
+}
+
+
+inline X509* LookforRoot(STACK_OF(X509)* sk) {
+  for (int i = 0; i < sk_X509_num(sk); i++) {
+    X509* cert = sk_X509_value(sk, i);
+    if (IsSelfSigned(cert))
+      return cert;
+  }
+  return nullptr;
+}
+
+
+// Whitelist check for certs issued by CNNIC. See
+// https://blog.mozilla.org/security/2015/04/02/distrusting-new-cnnic-certificates/
+bool IsWhitelistedServerCertificate(SSL* ssl) {
+  X509* root_cert;
+  X509_NAME* root_name;
+  unsigned char hash[CNNIC_WHITELIST_HASH_LEN];
+  unsigned int hashlen = CNNIC_WHITELIST_HASH_LEN;
+
+  if (SSL_is_server(ssl))
+    return true;
+
+  STACK_OF(X509)* chain = SSL_get_peer_cert_chain(ssl);
+
+  // Take the last cert as root at the first time.
+  root_cert = sk_X509_value(chain, sk_X509_num(chain)-1);
```

Comment:
What if the chain is empty? If that can never happen, can you add a `CHECK_GT(sk_X509_num(chain), 0);`?